manta-util/src/collections/vec_deque.rs

```diff
@@ -0,0 +1,175 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A double-ended queue (deque) implemented with a growable ring buffer.
+
+use crate::{array::BoxArray, collections::VecDeque};
+
+#[cfg(all(feature = "serde-alloc", feature = "serde-array"))]
+use crate::serde::{Deserialize, Serialize};
+
+#[doc(inline)]
+pub use alloc::collections::vec_deque::*;
+
+/// Multi-[`VecDeque`]
+#[cfg_attr(
+    all(feature = "serde-alloc", feature = "serde-array"),
+    derive(Deserialize, Serialize),
+    serde(
+        bound(deserialize = "T: Deserialize<'de>", serialize = "T: Serialize"),
+        crate = "crate::serde",
+        deny_unknown_fields
+    )
+)]
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct MultiVecDeque<T, const N: usize>(BoxArray<VecDeque<T>, N>);
+
+impl<T, const N: usize> MultiVecDeque<T, N> {
+    /// Builds a new empty [`MultiVecDeque`].
+    #[inline]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Returns a shared reference to the [`VecDeque`] at the given `level`.
+    #[inline]
+    pub fn at_level(&self, level: usize) -> &VecDeque<T> {
+        &self.0[level]
+    }
+
+    /// Returns a mutable reference to the [`VecDeque`] at the given `level`.
+    #[inline]
+    pub fn at_level_mut(&mut self, level: usize) -> &mut VecDeque<T> {
+        &mut self.0[level]
+    }
+
+    /// Returns the total number of elements in `self`.
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.0.iter().map(VecDeque::len).sum()
+    }
+
+    /// Returns `true` if `self` has no elements.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.0.iter().all(VecDeque::is_empty)
+    }
+
+    /// Returns a shared reference to the element at the given `index` in `self`.
```

Comment:
Why use a single `index` instead of a multi-index (`level` + `index`)? Much more intuitive in my opinion.

Reply:
This function is there to get a specific `index` in the queue when you don't know it's level. You can get the element by level with `self.at_level(level).get(index)`.